src/mscorlib/shared/Interop/Unix/System.Native/Interop.GetRandomBytes.cs

```diff
@@ -11,11 +11,11 @@ internal partial class Interop
     internal unsafe partial class Sys
     {
         [DllImport(Interop.Libraries.SystemNative, EntryPoint = "SystemNative_GetNonCryptographicallySecureRandomBytes")]
-        internal static extern unsafe void GetNonCryptographicallySecureRandomBytes(byte* buffer, int length);
+        internal static extern void GetNonCryptographicallySecureRandomBytes(in byte buffer, int length);
```

Comment:
This should use `byte*` or `ref`; not `in`.

Reply:
>The in modifier on parameters, to specify that an argument is passed by reference but not modified by the called method.

@jkotas `in` doesn't make sense on pinvoke?!I mean if put `in` i want pass valuetype by ref but avoid modification, i did a mistake, but that `in` wasn't honored...i get modification, what i'm missing?TYVM for patience.

Reply:
The buffer *is* modified by the called method in this case. So `in` is not appropriate for it.

Reply:
>The buffer is modified by the called method in this case. So in is not appropriate for it.

when you say "appropriate" do you mean that is misleading for "code reader"?Because umanaged code will ignore `in`?

Reply:
> in my test buffer was modified even with in on signature.

You have stated that you are not going to modify it, but you have modified it anyway. It happened to work for your test case.

It is similar to writing beyond end of the buffer. It also happens to work in most cases ... until there is something important beyond the end of the buffer and things break badly.

Reply:
It all clear now!Thank's a lot!

Reply:
> My personal preference for PInvoke declations is to use the unmanaged pointers (e.g. char*), not byrefs (e.g. ref char). I think unmanaged pointers are more clearly expressing the intent, and they also tent to be a bit faster because of char* is blittable type and so it avoids extra frame of auto-generated marshalling code.

I'm on the fence. Using `ref char` is nice in that it pins for you and avoids introducing `unsafe`. I'm inclined to think that's worth the extra frame. I can live with either approach, but I'd like to see us have clear guidance on this. @stephentoub - You had some feelings/thoughts on this didn't you?

Reply:
> Using `ref char` is nice in that it pins for you and avoids introducing `unsafe`

But it makes you use `MemoryMarshal.GetReference` that is more unsafe/dangerous than `unsafe` keyword.

Reply:
> But it makes you use MemoryMarshal.GetReference that is more unsafe/dangerous than unsafe keyword.

Why is that _more_ dangerous?

Reply:
Because of you have to be a rocket scientists to use `ref` as pointer correctly. Unmanaged pointers are much easier to reason about - they work just like in C/C++.

Reply:
> Because of you have to be a rocket scientists to use ref as pointer correctly.

How is it harder to use correctly? It seems to be a pretty straight-forward swap (`char*` => `ref char`). I'd much rather see this

``` C#
Span<byte> key = stackalloc byte[KeyLength];
Interop.GetRandomBytes(MemoryMarshal.GetReference(key), key.Length);
```
than this

``` C#
byte* pKey = stackalloc byte[KeyLength];
Interop.GetRandomBytes(pKey, KeyLength);
```

And even better I'd prefer to see wrappers that just take the span:

```
Span<byte> key = stackalloc byte[KeyLength];
Interop.GetRandomBytes(key);
```

If the internals/privates of GetRandomBytes P/Invoke with `char*` I don't care quite as much. I just really prefer to keep length context with the data as it is passed around.

Reply:
#14460